src/work/Work.h

```diff
@@ -16,125 +17,94 @@ namespace stellar
-    virtual size_t getMaxRetries() const;
-    uint64_t getRetryETA() const;
-
-    // Customize work behavior via these callbacks. onReset is called
-    // before any work starts (on addition, or retry). onStart is called
-    // when transitioning from WORK_PENDING -> WORK_RUNNING; onRun is
-    // called when run either from PENDING state _or_ after a SUCCESS
-    // callback reschedules running (see below). onFailure is only called
-    // on a failure before retrying; usually you can ignore it.
-    virtual void onReset();
-    virtual void onStart();
-    virtual void onRun();
-    virtual void onFailureRetry();
-    virtual void onFailureRaise();
-
-    // onSuccess is a little different than the others: it's called on
-    // WORK_SUCCESS, but it also returns the next sate desired: if you want
-    // to restart or keep going, return WORK_PENDING or WORK_RUNNING
-    // (respectively) from onSuccess and you'll be rescheduled to run more.
-    // If you want to force failure (and reset / retry) you can return
-    // WORK_FAILURE_RETRY or WORK_FAILURE_RAISE. After a retry count is
-    // passed, WORK_FAILURE_RETRY means WORK_FAILURE_RAISE anyways.
-    // WORK_FAILURE_FATAL is equivalent to WORK_FAILURE_RAISE passed up in
-    // the work chain - when WORK_FAILURE_FATAL is raised in one work item,
-    // all work items that leaded to this one will also fail without retrying.
-    virtual State onSuccess();
-
-    static std::string stateName(State st);
-    State getState() const;
-    bool isDone() const;
-    void advance();
-    void reset();
+    void shutdown() override;
 
   protected:
-    std::weak_ptr<WorkParent> mParent;
-    std::string mUniqueName;
-    size_t mMaxRetries{RETRY_A_FEW};
-    size_t mRetries{0};
-    State mState{WORK_PENDING};
-    bool mScheduled{false};
-
-    std::unique_ptr<VirtualTimer> mRetryTimer;
-
-    std::function<void(asio::error_code const& ec)> callComplete();
-    void run();
-    void complete(CompleteResult result);
-    void scheduleComplete(CompleteResult result = WORK_COMPLETE_OK);
-    void scheduleRetry();
-    void scheduleRun();
-    void
-    scheduleSuccess()
-    {
-        scheduleComplete();
-    }
-    void
-    scheduleFailure()
-    {
-        scheduleComplete(WORK_COMPLETE_FAILURE);
-    }
-    void
-    scheduleFatalFailure()
+    Work(Application& app, std::string name,
+         size_t retries = BasicWork::RETRY_A_FEW);
+
+    // Note: `shared_from_this` assumes there exists a shared_ptr to the
+    // references class. This relates to who owns what in Work interface.
+    // Thus, `addWork` should be used to create Work (then the parent holds
+    // the reference).
+    template <typename T, typename... Args>
+    std::shared_ptr<T>
+    addWork(Args&&... args)
     {
-        scheduleComplete(WORK_COMPLETE_FATAL);
+        auto child = std::make_shared<T>(mApp, std::forward<Args>(args)...);
+        addChild(child);
+        child->startWork(wakeUpCallback());
```

Comment:
Is that ever going to change for a single instance of Work? Maybe it can be moved into constructor parameter of Work class?

Reply:
I assumed it might, so kept it like this for flexibility.